src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -143,151 +131,77 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
-            {
-                // All the TryAddReference calls below must succeed as we're called from a caller that is pinning
-                // solution anyways with it's own refcount of at least 1.
-
-                solutionsToDispose.Add(_lastRequestedAnyBranchSolution.refCountedSolution);
-                _lastRequestedAnyBranchSolution = (solutionChecksum, solution.TryAddReference() ?? throw ExceptionUtilities.Unreachable);
-
-                if (fromPrimaryBranch)
-                {
-                    solutionsToDispose.Add(_lastRequestedPrimaryBranchSolution.refCountedSolution);
-                    _lastRequestedPrimaryBranchSolution = (solutionChecksum, solution.TryAddReference() ?? throw ExceptionUtilities.Unreachable);
-                }
-            }
-
-            foreach (var solutionToDispose in solutionsToDispose)
+            // Always try to retrieve cached solutions from the primary branch first.  That way we can use the solutions
+            // that were the real solutions of this Workspace, and not ones forked off from that.  This gives the
+            // highest likelihood of sharing data and being able to reuse caches and services shared among all
+            // components.
+            var primaryBranchRefCountedSolution = await _primaryBranchSolutionCache.TryFastGetSolutionAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+            if (primaryBranchRefCountedSolution != null)
+                return primaryBranchRefCountedSolution;
+
+            // Otherwise, have the any-branch solution try to retrieve or create the solution.
+            var anyBranchRefCountedSolution =
+                await _anyBranchSolutionCache.TryFastGetSolutionAsync(solutionChecksum, cancellationToken).ConfigureAwait(false) ??
+                await _anyBranchSolutionCache.SlowGetOrCreateSolutionAsync(
+                    solutionChecksum,
+                    cancellationToken => ComputeSolutionAsync(assetProvider, solutionChecksum, cancellationToken),
+                    cancellationToken).ConfigureAwait(false);
+
+            if (!updatePrimaryBranch)
             {
-                if (solutionToDispose != null)
-                    await solutionToDispose.DisposeAsync().ConfigureAwait(false);
+                // if we aren't updating the primary branch we're done and can just return the any-branch solution
+                return anyBranchRefCountedSolution;
             }
-        }
 
-        private async ValueTask<ReferenceCountedDisposable<LazySolution>> SlowGetSolutionAsync(
-            AssetProvider assetProvider,
-            Checksum solutionChecksum,
-            CancellationToken cancellationToken)
-        {
-            var currentSolution = this.CurrentSolution;
+            // We were asked to update the primary-branch solution.  So take the any-branch solution and promote it to
+            // the primary-branch-level.
```

Comment:
this is the clever bit.  we effectively ensure that whatever solution we made for this checksum gets promoted to be in the _primarySolutionCache.

What's critical to understand here is that each cache may have teh same checksum point to a *different* solution.  The _anyBranch cache can point from that checksum to a 'forked solution' that it generated (which then has some random branch id).

The _primaryBranch cache will then instead map from the checksum to an actual primary-branch-id solution if possible (which may be a different instance than the forked solution).

Later lookups always look in the _primaryBranch cache first, so this better solution will always be found.

This was *much* simpler to code up and ensure proper working versus trying to have a single mapping from `checksum->solution` that then might have to mutate in order update what that checksum pointed at.